File: backend/app/schemas/profile_visibility.py
"""Donor-profile element registry — the single source of truth for which
elements of the donor-facing child profile can be shown or hidden.

The visibility map (``orphans.profile_visibility``) mirrors the proven report
``section_visibility`` semantics exactly: an element is shown to donors UNLESS
its key is explicitly ``False``, so the default ``{}`` means "all elements
visible". The identity/header block is implicit and ALWAYS shown — it is NOT a
member of :class:`ProfileElement` and can never be hidden.

Adding a future element (``media``, ``father_memorial``, …) is a one-line change
here plus the matching block on the donor composition.
"""

from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel


class ProfileElement(StrEnum):
    """Every toggleable element of the donor child profile, keyed exactly to the
    data that exists today. Extend by adding one member (and its block on
    :class:`app.schemas.sponsored_profile.SponsoredOrphanProfile`)."""

    dream = "dream"
    her_world = "her_world"
    quran_growth = "quran_growth"
    multidim_growth = "multidim_growth"
    milestones = "milestones"
    recent_updates = "recent_updates"
    supervisor_word = "supervisor_word"
    since_you_began = "since_you_began"
    media = "media"
    in_her_words = "in_her_words"
    father_memory = "father_memory"
    health = "health"


def is_visible(profile_visibility: dict[str, bool], key: ProfileElement) -> bool:
    """An element is visible UNLESS its key is explicitly ``False`` in the stored
    map. An absent key (the default) ⇒ visible. Mirrors the report
    ``_visible_section`` rule so the two surfaces behave identically."""
    return profile_visibility.get(key.value) is not False


class ProfileElementState(BaseModel):
    """One row of the staff registry: an element + its current effective state."""

    key: ProfileElement
    visible: bool


class ProfileVisibilityRegistry(BaseModel):
    """Staff GET response: the full registry (every :class:`ProfileElement` with
    its current effective state) plus the raw stored map for transparency."""

    elements: list[ProfileElementState]
    stored: dict[str, bool]


class ProfileVisibilityUpdate(BaseModel):
    """Staff PUT body. Typing the map as ``dict[ProfileElement, bool]`` makes
    Pydantic reject (422) any unknown key or non-bool value at the boundary —
    no hand-rolled validation needed."""

    visibility: dict[ProfileElement, bool]

File: frontend/src/lib/profileVisibility.ts
import type { components } from "@/api/schema.gen";

import { api } from "./api";

// Donor-profile visibility types — mirror schema.gen.ts exactly (PR-1 backend).
// A ProfileElement is shown to the donor UNLESS its key is explicitly `false`;
// the default ({}) is "all visible". Hiding is enforced server-side — a hidden
// element never leaves the server — so this is not cosmetic.
export type ProfileElement = components["schemas"]["ProfileElement"];
export type ProfileElementState = components["schemas"]["ProfileElementState"];
export type ProfileVisibilityRegistry = components["schemas"]["ProfileVisibilityRegistry"];

// The full set of toggleable elements, in the donor-page display order. The
// identity/header block is intentionally absent — it is implicit and always
// shown, never a ProfileElement.
export const PROFILE_ELEMENTS: readonly ProfileElement[] = [
  "dream",
  "her_world",
  "quran_growth",
  "multidim_growth",
  "milestones",
  "recent_updates",
  "supervisor_word",
  "since_you_began",
  "in_her_words",
  "father_memory",
  "health",
  "home",
  "guardian",
  "siblings",
  "skills",
  "independence",
] as const;

/** Staff read of the donor-profile visibility registry for one orphan. */
export async function getProfileVisibility(orphanId: string): Promise<ProfileVisibilityRegistry> {
  const { data } = await api.get<ProfileVisibilityRegistry>(
    `/orphans/${orphanId}/profile-visibility`,
  );
  return data;
}

/** Staff write (replace semantics): send every ProfileElement key explicitly.
 * Returns the recomputed registry. */
export async function setProfileVisibility(
  orphanId: string,
  visibility: Record<ProfileElement, boolean>,
): Promise<ProfileVisibilityRegistry> {
  const { data } = await api.put<ProfileVisibilityRegistry>(
    `/orphans/${orphanId}/profile-visibility`,
    { visibility },
  );
  return data;
}
